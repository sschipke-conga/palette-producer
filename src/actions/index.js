

export const setUser = user => ({
  type: 'SET_USER',
  user
});

export const setAllProjects = allProjects => ({
  type: 'SET_PROJECTS',
  allProjects
});

export const setCurrentPalette = palette => ({
  type: 'SET_CURRENT_PALETTE',
  palette
});

export const setCurrentPaletteInfo = info => ({
  type: 'SET_CURRENT_PALETTE_INFO',
  info
});

export const setCurrentProjectInfo = info => ({
  type: 'SET_CURRENT_PROJECT_INFO',
  info
});

export const setAllPalettes = allPalettes => ({
  type: 'SET_PALETTES',
  allPalettes
});

export const toggleMenu = () => ({
  type: 'TOGGLE_MENU'
})

export const removePalette = (id) => ({
  type: 'REMOVE_PALETTE',
  id: parseInt(id)
});