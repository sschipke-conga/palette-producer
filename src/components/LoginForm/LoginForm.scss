.form-label-input-div {
  align-items: flex-start;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  height: 60px;
  margin-top: 20px;
  width: 60%;
  input {
    height: 100%;
    width: 100%;
  }
}