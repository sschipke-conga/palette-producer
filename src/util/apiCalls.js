export const createNewUser = async user => {
  let url = `${process.env.REACT_APP_PALETTE_PRODUCER_BACKEND_BASE_URL}/api/v1/signup`;
  let options = {
    method: "POST",
    body: JSON.stringify(user),
    headers: {
      "Content-Type": "application/json"
    }
  };
  let res = await fetch(url, options);

  if (res.status === 500) {
    throw Error("This username already taken");
  }

  if (!res.ok) {
    throw Error("Woops! Something went wrong");
  }

  return res.json();
};

export const loginUser = async user => {
  let url = `${process.env.REACT_APP_PALETTE_PRODUCER_BACKEND_BASE_URL}/api/v1/login`;
  let options = {
    method: "POST",
    body: JSON.stringify(user),
    headers: {
      "Content-Type": "application/json"
    }
  };
  let res = await fetch(url, options);

  if (res.status === 401) {
    throw Error("Username or password is incorrect");
  }

  if (!res.ok) {
    throw Error("Woops! Something went wrong");
  }

  return res.json();
};
