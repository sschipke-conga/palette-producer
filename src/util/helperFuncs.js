export const generateRandomHex = () => {
  let hexCode = "#" + Math.floor(Math.random() * 16777215).toString(16);
  if (hexCode.length < 7) {
    return generateRandomHex();
  }
  return hexCode;
};
