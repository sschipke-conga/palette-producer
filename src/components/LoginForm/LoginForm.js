import React, {Component} from 'react';
import {Redirect} from 'react-router-dom';
import {loginUser} from '../../util/apiCalls';
import './LoginForm.scss';

class LoginForm extends Component {
  constructor() {
    super();
    this.state = {
      username: "",
      password: "",
      error: "",
      isLoggedIn: false
    };
  }

  handleChange = (e) => {
    this.setState({ [e.target.id]: e.target.value });
  }

  handleSubmit = async (e) => {
    e.preventDefault()
    const {username, password, error
    } = this.state
    const user = {
      username,
      password
    }

    try {
      let res = await loginUser(user)
      this.setState({username:"", password:"", error:"", isLoggedIn: true})
      localStorage.setItem("user", JSON.stringify({user_id: res.id, username: res.username}))
      this.props.loadProjects(res.id)
    } catch ({message}) {this.setState({error: message})}
  }

  render() {
    const {username, password, error, isLoggedIn} = this.state;
    if(isLoggedIn) {
      return <Redirect to="/" />
    }
    let loginErrrorClass = error ? "input-error" : "";
    return (
      <form onSubmit={this.handleSubmit}>
        {error ? (<p className="error-message">{error}</p>) : (<p className="form-message">Good to see you, welcome back!</p>)}
        <div className="form-label-input-div">
          <label  htmlFor="username">Username
          </label>
          <input
            type="text"
            className={loginErrrorClass}
            placeholder="Enter your username"
            id="username"
            required
            value={username}
            onChange={this.handleChange}
          />
        </div>
        <div className="form-label-input-div">
          <label htmlFor="password">Password
          </label>
          <input
            type="password"
            className={loginErrrorClass}
            placeholder="Enter your password"
            id="password"
            minLength="8"
            required
            value={password}
            onChange={this.handleChange}
          />
        </div>
        <button type="submit">Login</button>
      </form>
    );
  }
}

export default LoginForm;