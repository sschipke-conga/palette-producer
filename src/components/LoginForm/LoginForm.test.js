import React from 'react';
import {LoginForm, mapDispatchToProps, mapStateToProps} from './LoginForm';
import { shallow } from 'enzyme';
import { loginUser } from '../../util/apiCalls';
import {setUser} from '../../actions/index'

jest.mock('../../util/apiCalls');

loginUser.mockImplementation(() => Promise.resolve({
  id: 5,
  name: 'Steve',
  password: '1234'
}));

describe('LoginForm', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<LoginForm
    setUser={jest.fn()}
    loadProjects={jest.fn()}
    />)
  })
  it('should match the initial snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  });
  describe('handleChange', () => {
    it('should update state if a user types in their username', () => {
      const mockEvent = { target: { id: 'username', value: 'Steve' } };
      expect(wrapper.state('username')).toEqual('');
      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.state('username')).toEqual(mockEvent.target.value)
    })
    it('should update state if a user types in their password', () => {
      const mockEvent = { target: { id: 'password', value: 'password' } };
      wrapper.instance().setState({ username: 'Steve' });
      expect(wrapper.state('username')).toEqual('Steve')

      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.state('password')).toEqual(mockEvent.target.value);
      expect(wrapper.state('username')).toEqual('Steve');
    })
    it('should run handleChange when the inputs detect a change', () => {
      const mockNameEvent = { target: { id: 'username', value: 'Robbie' } };
      const mockPasswordEvent = { target: { id: 'password', value: 'password' } };
      wrapper.instance().handleChange = jest.fn();
      wrapper.instance().forceUpdate();
      wrapper.find('[id="username"]').simulate('change', mockNameEvent);
      wrapper.find('[id="password"]').simulate('change', mockPasswordEvent);
      expect(wrapper.instance().handleChange).toHaveBeenCalledWith(mockNameEvent);
      expect(wrapper.instance().handleChange).toHaveBeenCalledWith(mockPasswordEvent);
    })
  })
  describe('handleSubmit', () => {
    let wrapper;
    const mockLoadProjects = jest.fn()
    beforeEach(() => {
      wrapper = shallow(<LoginForm 
      loadProjects={mockLoadProjects}
        setUser={jest.fn()}
      />)
    })
    let mockUser = {
      id: 4,
      username: "Susan",
      password: "password"
    }
    it('should prevent the default action when the form is submitted', () => {
      const mockEvent = { preventDefault: jest.fn() };
      wrapper.find('form').simulate('submit', mockEvent);
      expect(mockEvent.preventDefault).toHaveBeenCalled()
    })
    it('should call loginUser when called', () => {
      let mockNewUser = {
        username: "Susan",
        password: "password"
      }
      let mockEvent = { preventDefault: jest.fn() }
      loginUser.mockImplementation(() => {
        return Promise.resolve(mockUser)
      })
      wrapper.instance().setState({ username: 'Susan', password: 'password' });
      wrapper.instance().handleSubmit(mockEvent);
      expect(loginUser).toHaveBeenCalledWith(mockNewUser);
    })
    it('should set an error message to state if something goes wrong', async () => {
      let mockEvent = { preventDefault: jest.fn() }
      loginUser.mockImplementation(() => {
        throw Error('Woops')
      });
      wrapper.instance().forceUpdate()
      await wrapper.instance().handleSubmit(mockEvent);
      expect(wrapper.state('error')).toEqual('Woops')
    })
  })
  describe('alt snapShot', () => {
    it('should match the snapshot if a user is logged in', () => {
      let wrapper = shallow(<LoginForm 
        setUser={jest.fn()}
        loadProjects={jest.fn()}
      />);
      wrapper.instance().setState({ isLoggedIn: true });
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('mapStateToProps/mapDispatchToProps', () => {
    it('mapStateToProps gives all the movies in state', () => {
      const mockUser = {username: 'Dave', id: 4}
      const mockState = {
        user: mockUser
      };
      const expected = {
        user: mockUser
      };
      const mappedProps = mapStateToProps(mockState)
      expect(mappedProps).toEqual(expected)
    });

    it('calls dispatch with setAllPallettes action when it is called', () => {
      const mockUser = { username: 'Dave', id: 4 }
      const mockDispatch = jest.fn();
      const actionToDispatch = setUser('SET_USER', mockUser);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setUser('SET_USER', mockUser);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  })
})

