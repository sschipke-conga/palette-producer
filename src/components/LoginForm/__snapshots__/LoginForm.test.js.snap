// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm alt snapShot should match the snapshot if a user is logged in 1`] = `
<Redirect
  to="/"
/>
`;

exports[`LoginForm should match the initial snapshot 1`] = `undefined`;
